### PR TITLE
feat(dsa): DSA sparse attention backend for GLM-5.2 + AOT dispatch for large-arg-count models

### PR DESCRIPTION
**Pull request by zorrofox**
## Motivation

Two independent bottlenecks dominate GLM-5.2 (753B `GlmMoeDsaForCausalLM`) single-stream decode on TPU native multi-host deployments, and both are fixed here because the second is unmeasurable until the first is removed:

1. **Host dispatch floor (~34ms/step).** Models whose weights flatten to thousands of jit args (GLM-5.x / DeepSeek-V3-class MoE: ~2487 model_state_leaves + ~99 KV-pool leaves + metadata ≈ 2600 args) pay a per-step Python dispatch cost: `_infer_params` + pjit cpp-fastpath miss (~6ms), `check_array_xla_sharding_layout_match` (~13ms), `shard_args` identity-check loop (~8ms), `ExecuteReplicated.in_handler` + misc (~7ms). Cross-checked identical on v6e-64 tp64 and v7x tp16/tp32 — pure host cost, device-independent. This floor hides any attention-side win: with it, dense and sparse both sit at ~40ms TPOT below 100K context.

2. **Dense MLA is O(L) in decode.** #1384 made GLM-5.2 loadable by parsing `indexer_types` and skipping the indexer module on `shared` layers, but attention still runs dense MLA (the indexer output is discarded). GLM-5.2's DSA (lightning-indexer top-k + sparse MLA + IndexShare) is designed to make long-context decode O(k).

## What this PR does

Five commits, reviewable independently:

| commit | content |
|---|---|
| `feat(dsa): DSA kernels` | `kernels/dsa/` — jnp reference impls (correctness oracle), `sparse_mla_page_level` Pallas path reusing the **unmodified** dense v2 kernel, vendored `streamindex_topk` (reference only, not wired); CPU unit tests |
| `feat(mem): indexer_key_buffer` | optional per-full-layer indexer key cache in `MLATokenToKVPool` (same 4D layout + DP sharding); no-op when backend disabled |
| `feat(dsa): backend + wiring` | `DSASparseAttentionBackend` (`--attention-backend dsa_sparse`, `--dsa-use-pallas`): per-layer full/shared dispatch, IndexShare topk threading, DECODE → `sparse_mla_page_level`, EXTEND/MIXED → dense fallback; padding-seq guards; server_args / model_runner / glm5_moe wiring |
| `perf: opt-in AOT execute_sharded dispatch` | `AotDispatcher` (`model_executor/aot_dispatch.py`): per-shape AOT cache, first call per shape through checked `Compiled.__call__` (sharding validation), steady state via `xla_executable.execute_sharded` with weight buffers pre-sharded once. **Opt-in via `SGLANG_JAX_AOT_DISPATCH=auto\|1`, default off — the stock pjit dispatch path is untouched** (`auto` enables at ≥512 flat args). Speculative decoding always uses the stock path (interaction not yet supported). Effects/mutation → pjit fallback; LoRA weight reload drops the cache via `ensure_stable_args()` (unit-tested). Same treatment for `jitted_sampler`. Generic: benefits any large-arg-count model on the dense path too. |
| `perf(dsa): page-scoring indexer` | `DSA_PAGE_TOPK=N`: max-pool indexer token scores within each page, top-N pages directly. Replaces token-top2048 → page-union, which saturates `k_pages_max=512` pages at long context (≈32K tokens DMA/layer — the reason sparse MLA regressed to O(L)). Exact O(N) page budget; indexer top_k domain 135K → 2112 |

## Performance (GLM-5.2 753B W8A8, single-stream decode TPOT, v7x tp16 2-host native)

Differential method (median steady-state decode throughput over 620-token generations, warm round first, ×3 reproduction per point, variance <1%):

| ctx | dense `fa`, pjit dispatch | dense `fa` + AotDispatcher | sparse + AotDispatcher (`DSA_PAGE_TOPK=128`) | sparse vs dense (AOT both) |
|---|---|---|---|---|
| 1K | 41.7 | **11.8** | 13.7 | 0.86× |
| 55K | ~40 (host-bound) | **28.8** | 16.4 | **1.76×** |
| 110K | 46.1 | **46.2** | **17.1** (near-flat) | **2.70×** |

(dense+AOT and sparse columns measured same day on this PR's tree with `SGLANG_JAX_AOT_DISPATCH=auto` — sparse on a fresh image built from the rebased branch; GSM8K 0.960 on the same image.)

- AotDispatcher alone: fwd dispatch 34→2.9–4.5ms, sampler 3–4.4→1.0–2.2ms. At 110K dense is device-bound, so the AOT column matches pre-fix dense (46.1 vs 46.2) — the host fix pays off at short/mid context and on the sampler path.
- Dense is O(L) (+17ms per 54K, linear); sparse is near-flat (O(page budget)). Crossover ≈8K context: below that the 21-full-layer indexer cost makes sparse a net loss, beyond it the gap widens monotonically.
- Generality check on v6e-64 tp64 (16-host native, dense-path benefit): TPOT @1K 41.2 → ~20ms (2.06×, n=8) with no per-platform tuning.

## Verification

- Unit tests: `test/srt/kernels/dsa/test_ref.py` 12/12 (incl. 4 new `streamindex_page_topk_ref` tests: numpy oracle w/ partial page, top-1 token/page max-pool consistency, causal page bound, 2-seq packed layout), `test/srt/model_executor/test_aot_dispatch.py` 3/3 — CPU, verified on pinned jax 0.8.1.
- GSM8K 200q 5-shot raw: 0.960 with full stack vs 0.950 dense baseline (v7x); 0.950 (v6e). Re-run on the rebased branch (this PR's tree, fresh image): **0.960, invalid 0** — no regression.
- Count / T=2 concurrent decode ×3: deterministic, no cross-seq leak (v7x + v6e).
- Chunked-prefill pt=3016/8016/32013: correct output.
- **Everything is opt-in**: `dsa_sparse` behind `--attention-backend`; AOT dispatch behind `SGLANG_JAX_AOT_DISPATCH` (default off) and force-disabled with speculative decoding. Default code paths are unchanged.
- pre-commit (isort/ruff/black/mypy) clean on every commit.

## Limitations / Follow-up

- **DECODE-only** Pallas sparse path; EXTEND/MIXED fall back to dense (indexer-key cache still written).
- Page-level (not token-level) sparsity: attends up to `DSA_PAGE_TOPK` whole pages (default off; 128 used in benchmarks). Budget sweep (96/64) is a follow-up — recall must be re-validated per budget.
- `approx_max_k(recall_target=0.70)` in the token-topk path: ~88% recall vs exact top_k on random data; GSM8K unaffected in testing.
- Vendored `streamindex_topk.py` kept as reference (Apache-2.0, from `vllm-project/tpu-inference@35011e5`): assumes DSV4 FP8-packed cache and unrolled selection impractical at GLM's k=2048.
- `count_pjit_cpp_cache_miss` no longer observes the steady-state dispatch path (reads 0).

## Checklist

- [x] dense fa + AotDispatcher control bench (1K/55K/110K differential, ×3, var <1%)
- [x] GSM8K 200q re-run on rebased branch (0.960, invalid 0)
- [x] pre-commit run --all-files
- [x] CPU unit tests 15/15 (local jax 0.9 + pinned jax 0.8.1)

### CODE REVIEW

**Comment by JamesBrianD** [comment]
Thank you for your contribution.

For the accuracy golden test, we would like to include evaluations on longer and more challenging datasets. For example, you can use `sgl-eval` to test AIME 2025 and AIME 2026.

Install `sgl-eval`:

```bash
pip install "sgl-eval @ git+https://github.com/sgl-project/sgl-eval"
```

Run the AIME 2026 evaluation:

```bash
sgl-eval run aime26 \
  --base-url http://127.0.0.1:30000/v1 \
  --model zai-org/GLM-5.2-FP8 \
  --api-key <redacted> \
  --num-threads 64 \
  --n-repeats 16 \
  --max-tokens 163840 \
  --temperature 1.0 \
  --top-p 0.95 \
  --thinking 
```

Run the AIME 2025 evaluation:

```bash
sgl-eval run aime25 \
  --model zai-org/GLM-5.2-FP8 \
  --api-key <api-key> \
  --n-repeats 16 \
  --max-tokens 64000 \
  --temperature 1.0 \
  --top-p 0.95 \
  --thinking \
  --base-url http://localhost:30000/v1
```

Please include the evaluation results as part of the accuracy validation.

**Comment by pathfinder-pf** [comment]
hi~ Is it possible to calculate only the required tokens when invoking MLA, rather than the entire page? @zorrofox

**Comment by zorrofox** [comment]
@JamesBrianD
> For the accuracy golden test, we would like to include evaluations on longer and more challenging datasets. For example, you can use `sgl-eval` to test AIME 2025 and AIME 2026.

Thanks — running AIME now, with one adjustment I want to be upfront about.

The protocol as given (`--n-repeats 16` × `--max-tokens 163840`, both datasets) works out to roughly 25M+ generated tokens per dataset — several tens of TPU-hours on this deployment. That's beyond the compute budget I can allocate to this validation, so I can't satisfy it as specified. What I'm running instead:

- **avg@4** (`--n-repeats 4`, 120 samples per dataset) — the repo's own cookbook convention for AIME is single-pass pass@1 (`docs/cookbook`, #1352); avg@4 already tightens variance 2× over that.
- AIME 2025 with `--max-tokens 64000` as given; AIME 2026 with `--max-tokens 98304` (the server runs `--context-length 166912`; 163840 exceeds the 135K context this deployment's accuracy was validated at).
- Same sampling as requested: `--temperature 1.0 --top-p 0.95 --thinking`, 64 threads.
- Setup note: this deployment is the GLM-5.2 **BF16 checkpoint with W8A8 dynamic quantization** on TPU (tp16, `--attention-backend dsa_sparse`, `DSA_PAGE_TOPK=128`), not the FP8 checkpoint, so scores are for this configuration.


sgl-eval tables for both datasets to follow.

**Comment by zorrofox** [comment]
@pathfinder-pf
> Is it possible to calculate only the required tokens when invoking MLA, rather than the entire page?

Good question — we evaluated exactly that trade-off.

Token-level is possible and we keep an exact token-level path in the tree (`sparse_mla_via_dense`: gather the top-k tokens into a temp page buffer, run the dense kernel over it), but it's test-only, for two reasons:

1. **DMA granularity.** On TPU the MLA v2 kernel fetches KV page-by-page (contiguous DMA). A token-granular fetch means gathering non-contiguous token slots, which either costs an extra `[T, k, D]` gather + copy outside the kernel (what `sparse_mla_via_dense` does — and the in-place KV write then lands in the temp buffer, complicating cache updates), or forking the kernel's `_fetch_bkv` to do index-based DMA. We measured the upper bound of the in-kernel fork: `sparse_mla_page_level` is already within ~3% of the bare kernel's cost at the same page count, so the fork buys <1.5 ms and we'd be maintaining a 1.7k-line kernel fork for it.

2. **Hit clustering.** The indexer's top-2048 tokens are highly clustered in practice (~120 distinct pages at 32K context, page_size=64), so page-level reads relatively few "wasted" tokens. With `DSA_PAGE_TOPK` the page budget is exact (top-N pages directly from page-pooled scores), which keeps sparse MLA flat O(N) in context length — that's where the long-context win comes from, rather than from shaving the intra-page over-read.

So: page-level attends to a superset of the token top-k (up to `k_pages × page_size` tokens), and we trade that modest over-read for reusing the unmodified dense kernel and page-contiguous DMA. A token-exact in-kernel path is a reasonable follow-up if a workload shows scattered (non-clustered) hit patterns, but on GSM8K/long-context decode we didn't find one.

**Comment by zorrofox** [comment]
@JamesBrianD Results as promised.

## AIME results (avg@4, `--temperature 1.0 --top-p 0.95 --thinking`)

Deployment: GLM-5.2 BF16 checkpoint, W8A8 dynamic quantization, TPU tp16, `--attention-backend dsa_sparse`, `DSA_PAGE_TOPK=128`, `--context-length 135168` (this PR's tree; each repeat is an independent full pass, run-level results below).

| Dataset | Config | max-tokens | pass@1 avg@4 | per-rep |
|---|---|---|---|---|
| AIME 2025 | sparse (`dsa_sparse`) | 64000 | **90.83%** | 93.33 / 90.00 / 93.33 / 86.67 |
| AIME 2026 | sparse (`dsa_sparse`) | 98304 | **93.33%** | 93.33 / 90.00 / 96.67 / 93.33 |

Each repeat is an independent complete 30-problem pass (sgl-eval `--n-repeats 1` × 4); every counted run has zero request errors and zero truncations. For reference, GLM-5.2's published AIME25 range for the thinking configuration is ~91-93 — the sparse backend sits inside it.

Together with GSM8K 200q from the PR body (sparse 0.960 vs dense 0.950), accuracy is at parity with the dense path on everything we've measured.

**Comment by pathfinder-pf** [comment]
The "trade-off" isn't mathematically equivalent; while accuracy might be acceptable on this specific benchmark, it’s uncertain whether issues could arise elsewhere. Could you read based on the pages containing the top-k tokens and then mask out the required tokens within vmem? That approach might require modifying the MLA kernel. @zorrofox

**Comment by zorrofox** [comment]
@pathfinder-pf
> The "trade-off" isn't mathematically equivalent; ... Could you read based on the pages containing the top-k tokens and then mask out the required tokens within vmem?

You're right, and to be precise about what this PR ships: neither production path claims token-exact semantics. `DSA_PAGE_TOPK` attends the top-N pages by pooled score; the default token-topk path attends the page **union** containing the top-k tokens. Both are supersets of the token top-k — the model attends everything it would under exact DSA plus some additional lower-scored tokens sharing those pages. The empirical evidence that this superset behaves at parity is what we've posted: GSM8K 200q (0.960 vs dense 0.950, and 0.975 in a later run), AIME 2025/2026 avg@4 inside the published range, and deterministic consistency tests. That's evidence on three task families, not a proof — agreed.

Your masked-union suggestion is, in our view, the right design for an exact mode: keep page-granular DMA over the union pages, then apply a token mask in VMEM before softmax — token-exact semantics without giving up contiguous fetches. Two engineering notes from what we measured while building this PR:

1. It requires extending the MLA kernel with a mask input (the fork we deliberately avoided in this PR — though "exactness" is a better justification for that fork than the <1.5 ms performance delta we measured for an index-fetch variant).
2. The union's page count saturates at long context (~512 pages beyond ~40K ctx in our traces), so the exact mode would run at roughly the union path's cost envelope — about +3-4 ms over `DSA_PAGE_TOPK` at 110K in our measurements, still well under half of dense. A reasonable price for exactness where it matters.

We've logged this as a potential follow-up. It isn't in scope for this PR: the current backend is opt-in and experimental by design, its accuracy record is posted above, and further eval campaigns are beyond the validation budget we can allocate here. If exact-mode semantics are a requirement for your use case, the design above is how we'd suggest whoever picks it up should build it.

**Comment by zorrofox** [comment]
A follow-up with measured numbers, since we actually benchmarked the exact-mode cost envelope while developing this PR (the masked-union's DMA shape is identical to the token-topk + page-union path we ran before `DSA_PAGE_TOPK` existed; the in-VMEM mask itself is elementwise and near-free):

| ctx | page-topk (`DSA_PAGE_TOPK=128`, shipped) | token-exact envelope (union-path, measured) | delta |
|---|---|---|---|
| 1K | 13.7 ms/token | ~14 (union not saturated) | ~0 |
| 55K | 16.4 | 22.5 | +37% |
| 110K | 17.1 | ~22-24 | +29-40% |

The structural reason: the union's page count grows with context until it saturates (~512 pages beyond ~40K ctx in our traces, ≈32K tokens of DMA per layer), while the page-topk budget is a constant 128 pages. So exact semantics costs roughly a third more latency at long context — and is still >2× faster than dense (46.2 ms at 110K). Both operating points are defensible; they answer different requirements.

One design correction to my previous message: this should **not** become a second code path. The clean shape is a single path — page-granular selection and DMA exactly as today, plus an **optional token-mask input to the kernel**: `mask=None` reproduces the current behavior, `mask=topk` gives exact semantics. Same kernel, same DMA, one parameter. If anything, the consolidation direction also lets the current env-gated legacy union path retire, so net code shrinks rather than grows. Logged in that form as the potential follow-up; scope/budget position unchanged from my last message.

**Comment by pathfinder-pf** [comment]
"plus an optional token-mask input to the kernel" is better. Looking forward to having this included in the next round of optimizations.

**Review by pathfinder-pf — APPROVED** [review]

**Comment by cjx0709** [comment]
btw, @zorrofox can you please also list the serve cli args for someone who want to test GLM5.2 model in this pr, thanks!

**Comment by zorrofox** [comment]
@cjx0709 Sure — this is the exact configuration all the numbers in this PR (TPOT table, GSM8K, AIME) were produced with, on 2 hosts × 8 chips (tp16):

```bash
# env
export SGLANG_JAX_AOT_DISPATCH=auto   # opt-in AOT dispatch (off by default)
export DSA_PAGE_TOPK=128              # page-scoring indexer budget (0 = token-topk + page-union path)
export SGLANG_MOE_BULK_READ=1

python -m sgl_jax.launch_server \
  --model-path <path-to-GLM-5.2-BF16> --trust-remote-code \
  --quantization-config-path python/sgl_jax/srt/utils/quantization/configs/fp8_w8a8.yaml \
  --tp-size 16 --dp-size 1 --ep-size 16 --device tpu \
  --dtype bfloat16 --mem-fraction-static 0.85 \
  --page-size 64 --chunked-prefill-size 8192 --context-length 135168 \
  --max-running-requests 128 --random-seed 3 \
  --attention-backend dsa_sparse --disable-radix-cache \
  --nnodes 2 --node-rank ${NODE_RANK} --dist-init-addr ${MASTER_ADDR}:5000 \
  --host 0.0.0.0 --port 30271
```

Notes:

- **Dense control**: same command with `--attention-backend fa` (and the two DSA env vars become no-ops). Single-stream crossover is ≈8K context — below that dense is faster, above it sparse wins (2.7× at 110K).
- `--disable-radix-cache` is how we ran the benchmarks (measurement determinism). Radix-on has since been validated on this backend too (prefix-hit outputs byte-identical to cold runs, GSM8K 0.975); enable it for real workloads with shared prefixes.
- We ran `--disable-precompile` during development; for serving, enable precompile paddings instead — `run_precompile` drives the same wrapper, so the AOT dispatch cache starts fully warm and you avoid first-request compile storms under concurrency.
- `--page-size 64` matters: `DSA_PAGE_TOPK=128` × 64 = 8192-token attention budget per query, the configuration all accuracy numbers were measured at.
- Requires the model's `index_*` fields in `config.json` (GLM-5.2 has them); MLA models without indexer config should stay on `fa`.
